### PR TITLE
Allow customizing ChatViewController used in Threads

### LINKED ISSUE #136: Different style for threads
## What are you trying to achieve?

Currently we are using Chat ontop of a video stream.. so we setup the message styles with clear backgrounds. 

However if replies is turned on for the ChannelType when you try and do a reply it does a push navigation and ends up showing the view controller of what brought you to the video stream.
This is also transparent.

What I would like is if there is a way to use not inherit those styles... ie if there was a way in `ChatViewController` to have a style for Threaded Messages

### PR DESCRIPTION
**Pull request by cardoso**
Made this based on how ChannelsViewController allows customizing the ChatViewController.

Closes #136

### CODE REVIEW

File: Sources/UI/Chat/Chat View Controller/ChatViewController.swift [hunk 1]
```diff
@@ -308,6 +308,19 @@ open class ChatViewController: ViewController, UITableViewDataSource, UITableVie
         defaultActionsContextMenu(from: cell, for: message, locationInView: locationInView)
     }
     
+    /// Creates a chat view controller for the message being replied to.
+    ///
+    /// - Parameters:
+    ///     - channelPresenter: the channel presenter of the message being replied to.
+    /// - Returns: a chat view controller.
+    open func createChatViewController(with channelPresenter: ChannelPresenter) -> ChatViewController {
```
Comment (b-onc) [inline Sources/UI/Chat/Chat View Controller/ChatViewController.swift:316]: Can you rename this to `createThreadChatViewController` or something along the lines of this? Having `createChatViewController` under `ChatViewController` is strange

File: Sources/UI/Chat/Chat View Controller/ChatViewController.swift [hunk 2]
```diff
@@ -308,6 +308,19 @@ open class ChatViewController: ViewController, UITableViewDataSource, UITableVie
         defaultActionsContextMenu(from: cell, for: message, locationInView: locationInView)
     }
     
+    /// Creates a chat view controller for the message being replied to.
```
Comment (b-onc) [inline Sources/UI/Chat/Chat View Controller/ChatViewController.swift:311]: Can you add a note about how one should override this to add custom styling to their threads?

File: Example/Sources/CustomChatViewController.swift [hunk 3]
```diff
@@ -66,6 +66,16 @@ class CustomChatViewController: ChatViewController {
             .disposed(by: disposeBag)
     }
     
+    override func createChatViewController(with channelPresenter: ChannelPresenter) -> ChatViewController {
+        let controller = super.createChatViewController(with: channelPresenter)
```
Comment (cardoso) [inline Example/Sources/CustomChatViewController.swift:70]:
```suggestion
    override func createThreadChatViewController(with channelPresenter: ChannelPresenter) -> ChatViewController {
        let controller = super.createThreadChatViewController(with: channelPresenter)
```

File: Sources/UI/Chat/Chat View Controller/ChatViewController+Cells.swift [hunk 4]
```diff
@@ -209,9 +209,7 @@ extension ChatViewController {
         messagePresenter.imageAttachmentExtraDataCallback = presenter.imageAttachmentExtraDataCallback
         messagePresenter.messagePreparationCallback = presenter.messagePreparationCallback
         
-        let chatViewController = ChatViewController(nibName: nil, bundle: nil)
-        chatViewController.style = style
-        chatViewController.presenter = messagePresenter
+        let chatViewController = createChatViewController(with: messagePresenter)
```
Comment (cardoso) [inline Sources/UI/Chat/Chat View Controller/ChatViewController+Cells.swift:212]:
```suggestion
        let chatViewController = createThreadChatViewController(with: messagePresenter)
```

File: Sources/UI/Chat/Chat View Controller/ChatViewController.swift (hunk 1 above)
Comment (cardoso) [inline Sources/UI/Chat/Chat View Controller/ChatViewController.swift:316]:
```suggestion
    /// Creates a chat view controller for the message being replied to. Override this to change style and other properties of the thread's view controller.
    ///
    /// - Parameters:
    ///     - channelPresenter: the channel presenter of the message being replied to.
    /// - Returns: a chat view controller.
    open func createThreadChatViewController(with channelPresenter: ChannelPresenter) -> ChatViewController {
```

File: Sources/UI/Chat/Chat View Controller/ChatViewController.swift [hunk 5]
```diff
@@ -308,6 +308,19 @@ open class ChatViewController: ViewController, UITableViewDataSource, UITableVie
         defaultActionsContextMenu(from: cell, for: message, locationInView: locationInView)
     }
     
+    /// Creates a chat view controller for the message being replied to. Override this to change style and other properties of the thread's view controller.
```
Comment (b-onc) [inline Sources/UI/Chat/Chat View Controller/ChatViewController.swift:311]: This should be on a different line, in discussion:
```swift
/// Creates a chat view ....
/// 
/// Override this to style...
```

**Comment by cardoso** [comment]
> I was thinking about adding a `threadStyle` but this is actually better!

Almost did that as well... Or even a call back `threadStyle(for message: Message)`. But figured this to be the most flexible approach.

File: Sources/UI/Chat/Chat View Controller/ChatViewController.swift [hunk 6]
```diff
@@ -308,6 +308,19 @@ open class ChatViewController: ViewController, UITableViewDataSource, UITableVie
...
+    ///
+    /// - Parameters:
```
Comment (cardoso) [inline Sources/UI/Chat/Chat View Controller/ChatViewController.swift:313]:
```suggestion
    /// Creates a chat view controller for the message being replied to.
    ///
    /// Override this to change style and other properties of the thread's view controller.
    /// - Parameters:
```

File: Sources/UI/Chat/Chat View Controller/ChatViewController.swift (hunk 5 above)
Comment (cardoso) [reply Sources/UI/Chat/Chat View Controller/ChatViewController.swift]: Done! Thanks

**Comment by b-onc** [comment]
LGTM! Can you make sure it passes Danger without error (except the tests warning)

**Comment by cardoso** [comment]
About the file length warning, I'm thinking about moving the TableView methods (from line 372 onwards) to a separate file "ChatViewController+TableView.swift" what do you think @b-onc ?

**Comment by b-onc** [comment]
@cardoso you can't since open functions in extensions cannot be overridden

> About the file length warning, I'm thinking about moving the TableView methods (from line 372 onwards) to a separate file "ChatViewController+TableView.swift" what do you think @b-onc ?

File: CHANGELOG.md [hunk 7]
```diff
@@ -43,6 +44,7 @@ open func updateFooterTypingUserAvatarView(footerView: ChatFooterView, user: Use
 - SPM support [#156](https://github.com/GetStream/stream-chat-swift/issues/156).
 - Made `SubscriptionBag.init` public [#172](https://github.com/GetStream/stream-chat-swift/issues/172).
 - Unused `RxBlocking` dependency removed [#177](https://github.com/GetStream/stream-chat-swift/pull/177).
+- Reconnection now automatically re-watches all channels watched up to that point [#178](https://github.com/GetStream/stream-chat-swift/pull/178).
```
Comment (cardoso) [inline CHANGELOG.md:47]:
```suggestion
```

**Pull request merged**